### PR TITLE
[FIX] crm, project: fixing some non-total order

### PR DESCRIPTION
**Pull request by Xavier-Do**
Non total order can lead to randomized order in some case, making some
test behaviour random and/or difficult to debug.

Adding some fallback 'id' order shouldn't hurt

### CODE REVIEW

**Comment by Xavier-Do** [comment]
@robodoo override=ci/style
Modifying existing line, not a big deal for a domain.

File: addons/crm/models/crm_lead.py [hunk 1]
```diff
@@ -1957,9 +1957,9 @@ def _pls_get_naive_bayes_probabilities(self, batch_mode=False):
                 if field == 'stage_id' and value in won_stage_ids:
                     won_leads.add(lead_id)
                 leads_fields.add(field)
-
+        leads_fields = sorted(leads_fields)
```
Comment (tde-banana-odoo) [inline addons/crm/models/crm_lead.py:1960]: Why is it necessary ? I don't see where queries would be impacted by the ordering ?

File: addons/crm/models/crm_lead.py [hunk 2]
```diff
@@ -2351,7 +2351,7 @@ def _pls_get_won_lost_total_count(self, team_results):
         :return: won count, lost count and total count for all records in frequencies
         """
         # TODO : check if we need to handle specific team_id stages [for lost count] (if first stage in sequence is team_specific)
-        first_stage_id = self.env['crm.stage'].search([('team_id', '=', False)], order='sequence', limit=1)
+        first_stage_id = self.env['crm.stage'].search([('team_id', '=', False)], order='sequence, id', limit=1)
```
Comment (tde-banana-odoo) [inline addons/crm/models/crm_lead.py:2354]: Same desc ?

File: addons/crm/models/crm_lead.py [hunk 3]
```diff
@@ -2365,7 +2365,7 @@ def _pls_prepare_frequencies(self, lead_values, leads_pls_fields, target_state=N
         pls_fields = leads_pls_fields.copy()
         frequencies = dict((field, {}) for field in pls_fields)
 
-        stage_ids = self.env['crm.stage'].search_read([], ['sequence', 'name', 'id'], order='sequence')
+        stage_ids = self.env['crm.stage'].search_read([], ['sequence', 'name', 'id'], order='sequence, id')
```
Comment (tde-banana-odoo) [inline addons/crm/models/crm_lead.py:2368]: Same desc ?

File: addons/crm/models/crm_lead.py [hunk 4]
```diff
@@ -2446,7 +2446,7 @@ def _pls_get_lead_pls_values(self, domain=[]):
             self.flush(['probability'])
             query = """SELECT id, probability, %s
                         FROM %s
-                        WHERE %s order by team_id asc"""
+                        WHERE %s order by team_id asc, id"""
```
Comment (tde-banana-odoo) [inline addons/crm/models/crm_lead.py:2449]: If we want to be coherent with the model, should be id desc :)

File: addons/crm/models/crm_lead.py [hunk 5]
```diff
@@ -1957,9 +1957,9 @@ def _pls_get_naive_bayes_probabilities(self, batch_mode=False):
                 if field == 'stage_id' and value in won_stage_ids:
                     won_leads.add(lead_id)
                 leads_fields.add(field)
-
+        leads_fields = sorted(leads_fields)
         # get all variable related records from frequency table, no matter the team_id
-        frequencies = self.env['crm.lead.scoring.frequency'].search([('variable', 'in', list(leads_fields))], order="team_id asc")
+        frequencies = self.env['crm.lead.scoring.frequency'].search([('variable', 'in', list(leads_fields))], order="team_id asc, id")
```
Comment (tde-banana-odoo) [inline addons/crm/models/crm_lead.py:1962]: Shouldn't we set this as model order (seems it is missing) instead ?

Comment (tde-banana-odoo) [reply addons/crm/models/crm_lead.py:1962]: And ``crm.lead.scoring.frequency.field`` has no order non plus

**Review by tde-banana-odoo — COMMENTED** [review]
Bisous

File: addons/crm/models/crm_lead.py (hunk 5 above)
Comment (Xavier-Do) [reply addons/crm/models/crm_lead.py:1962]: True, setting the order on the model may be a good idea, but I'm not sure this is wanted. My goal is to do transparent fix, not to  change the behavior without knowing the business needs :) 
If a model has no order the default order is 'id' so there is no problem in this case. 
A potential problem is to set on a model _order = 'non_unique_field', in this case a falback on id is a good idea.

File: addons/crm/models/crm_lead.py (hunk 2 above)
Comment (Xavier-Do) [reply addons/crm/models/crm_lead.py:2354]: I don't understand :)

File: addons/crm/models/crm_lead.py (hunk 4 above)
Comment (Xavier-Do) [reply addons/crm/models/crm_lead.py:2449]: We can change that!

**Review by tde-banana-odoo — APPROVED** [review]
Thanks, LGTM :) I think we can backport to 15.0, as this should not have any impact on functional side and on any method API. As it lessens non deterministic issues, I think 15 is a safe version.

**Comment by tde-banana-odoo** [comment]
Esta bien rouge :D

**Comment by Xavier-Do** [comment]
> Esta bien rouge :D

Un hook a été perdu, c'est fixé :)

**Comment by Xavier-Do** [comment]
@robodoo r+ (tde is okay)

**Comment by Xavier-Do** [comment]
you drunk robodoo just merge that!

**Comment by tde-banana-odoo** [comment]
Go home robodoo, you are drunk.

**Pull request closed without merging**